### PR TITLE
Port of badge.team terminal interface

### PR DESCRIPTION
**Pull request by MatthewWilkes**
This ports more of the badge.team functionality, especially the terminal interface.

### CODE REVIEW

**Comment by dpslwk** [comment]
upysh is in micropython-lib so can be frozen rather than copied in

**Comment by MatthewWilkes** [comment]
As this is working against the web view now, marking this as ready for review. Addressed @dpslwk's comment on upysh.

File: modules/dashboard/terminal/picture.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	apps = uos.listdir('/lib')
+	for app in apps:
+		files = uos.listdir('/lib/'+app)
+		for f in files:
+			if f.endswith('.png'):
+				pictures.append("/lib/"+app+"/"+f)
+except:
+	pass
+
+current = machine.nvs_getstr('splash', 'logo')
```
Comment (dpslwk) [inline modules/dashboard/terminal/picture.py:24]: didi someone saw this does not work on your µPy version?

Comment (tomsci) [reply modules/dashboard/terminal/picture.py:24]: correct that API has been replaced by a class `machine.NVS` or something like that. And we don't use nvs at all for any of our settings - and we don't have a config for setting a splashscreen either.

Comment (MatthewWilkes) [reply modules/dashboard/terminal/picture.py:24]: Yeah, I missed this code as it's not used in the set of terminal apps we're using. Will adjust it.

File: modules/system.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+from textwindow import TextWindow
+import tidal
+
+
+def showMessage(message):
+    window = TextWindow(bg=tidal.RED)
+    window.cls()
+    x = 0
+    while x < len(message):
```
Comment (tomsci) [inline modules/system.py:9]: You can use `TextWindow.flow_lines()` to do this, if you prefer.

File: modules/term_menu.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		items = ["File downloader", "< Return to main menu"]
+		callbacks = [self.opt_downloader, self.menu_main, self.menu_main]
+		cb = term.menu("Tools", items)
+		self.menu = callbacks[cb]
+	
+	def menu_safe(self):
+		items = ["Main menu"]
+		callbacks = [self.menu_main]
+		cb = term.menu("You have started the badge in safe mode!", items)
+		self.menu = callbacks[cb]
```
Comment (tomsci) [inline modules/term_menu.py:78]: `<pedant>`Is there a reason we've switched from spaces to tabs in some of these files?`</pedant>`

Comment (MatthewWilkes) [reply modules/term_menu.py:78]: Yeah, badge.team love tabs, I thought I'd fixed all of them!

**Comment by MatthewWilkes** [comment]
This isn't bug free (there are issues with reopening menus that were previously closed) but it enables the webserial interface so think it's a good basis for user programming. Any further review comments for changes would be welcomed, but think this is probably important for the initial distribution build.

File: modules/system.py [hunk 4]
```diff
@@ -5,10 +5,7 @@
 def showMessage(message):
     window = TextWindow(bg=tidal.RED)
     window.cls()
-    x = 0
-    while x < len(message):
-        window.println(message[x:x+16])
-        x += 16
+    window.flow_lines(message)
```
Comment (tomsci) [inline modules/system.py:8]: This doesn't actually print the text, it just splits it up into an array of strings the right length :)

I should probably add a print_lines API huh...

**Review by tomsci — CHANGES_REQUESTED** [review]
other than the flow_lines thing, looks good.

**Review by tomsci — APPROVED** [review]

**Comment by tomsci** [comment]
As discussed let's not leave this on a branch any longer. Merging.

**Pull request merged**